addon/mixins/disposable.js

```diff
@@ -0,0 +1,144 @@
+import Ember from 'ember';
+import getOrAllocate from '../utils/get-or-allocate';
+
+const { Mixin } = Ember;
+
+/**
+ DisposableMixin provides a mechanism register disposables with automatic disposing when the
+ host object is destroyed.
+
+ @class ContextBoundTasksMixin
+ @public
+ */
+export default Mixin.create({
+  init() {
+    this._super(...arguments);
+
+    this._registeredDisposables = undefined;
+  },
+
+  /**
+   Adds a new disposable to the object. A disposable is a function that
+   disposes of bindings that are outside of Ember's lifecyle. This essentially
+   means you can register a function that you want to run to automatically tear
+   down any bindings when the Ember object is destroyed.
+
+   Example:
+
+   ```js
+   // app/components/foo-bar.js
+   import Ember from 'ember';
+   import DOMish from 'some-external-library';
+
+   const { run } = Ember;
+
+   export default Component.extend({
+     init() {
+       this.DOMish = new DOMish();
+
+       this.bindEvents();
+     },
+
+     bindEvents() {
+       let onFoo = run.bind(this.respondToDomEvent);
+       this.DOMish.on('foo', onFoo);
+
+       this.domFooToken = this.registerDisposable(() => this.DOMish.off('foo', onFoo));
+     },
+
+     respondToDOMEvent() {
+       // do something
+     }
+   });
+   ```
+
+  @method registerDisposable
+  @param { Function } disposable
+  @returns A label representing the position of the disposable
+  @public
+  */
+  registerDisposable(disposable) {
+    if (typeof disposable !== 'function') {
+      throw new Error('You must pass a function as a disposable');
+    }
+
+    let disposables = getOrAllocate(this, '_registeredDisposables', []);
+
+    return disposables.push(disposable) - 1;
+  },
+
+  /**
+  Runs a disposable identified by the supplied token.
+
+ ```js
+  // app/components/foo-bar.js
+  import DOMish from 'some-external-library';
+  import Ember from 'ember';
+
+  const { run } = Ember;
+
+  export default Component.extend({
+    init() {
+      this.DOMish = new DOMish();
+
+      this.bindEvents();
+    },
+
+    bindEvents() {
+      let onFoo = run.bind(this.respondToDomEvent);
+      this.DOMish.on('foo', onFoo);
+
+      this.domFooToken = this.registerDisposable(() => this.DOMish.off('foo', onFoo));
+    },
+
+    respondToDOMEvent() {
+      // do something
+    },
+
+    actions: {
+      cancelDOM() {
+        this.runDisposable(this.domFooToken);
+      }
+    }
+  });
+  ```
+
+  @method runDisposable
+  @param {any} label
+  @public
+  */
+  runDisposable(token) {
+    runDisposable(this._registeredDisposables, token);
+  },
+
+  willDestroy() {
+    this._super(...arguments);
+
+    runDisposables(this._registeredDisposables);
```

Comment:
Generally speaking, the `willDestroy` should be the _inverse_ of setup work.  This means that the default pattern should be:

```js
willDestroy() {
  // do custom cleanup work

  // then call super afterwards
  this._super(...arguments);
}
```